pkg/provisioning/bg/validation.go

```diff
@@ -0,0 +1,562 @@
+	}
+	if res == true && err == nil {
+		t.Result = resultValid
+	} else if res == true && err != nil {
+		t.Result = resultWarning
+	}
+	return res
+}
+
+// Declaration of CBnTTests
+var (
+	hasFitEntries = CBnTTest{
+		Name:     "HasFitEntries",
+		function: HasValidFitEntries,
+	}
+	kmValidHeader = CBnTTest{
+		Name:     "KM Header",
+		function: KMStructureValidHeader,
+	}
+	kmValidKMSigOffset = CBnTTest{
+		Name:     "KM Signature Offset",
+		function: KMStructureValidKMSOffset,
+	}
+	kmValidReserverd2 = CBnTTest{
+		Name:     "KM Reserved2",
+		function: KMStructureValidReserved2,
+	}
+	kmrevisionvalid = CBnTTest{
+		Name:     "KM Revision",
+		function: KMStructureValidRevision,
+	}
+	kmsvnvalid = CBnTTest{
+		Name:     "KM SVN",
+		function: KMStructureValidSVN,
+	}
+	kmidvalid = CBnTTest{
+		Name:     "KM ID",
+		function: KMStructureValidID,
+	}
+	kmpubkeyhash = CBnTTest{
+		Name:     "KM Pub Key Hash Alg",
+		function: KMStructureValidKMPubHashAlg,
+	}
+	kmkeyhashcount = CBnTTest{
+		Name:     "KM Key Hash Count",
+		function: KMStructureValidKeyHashCount,
+	}
+	kmbpmkeyhash = CBnTTest{
+		Name:     "KM BPM Pub Key Hash",
+		function: KMStructureValidHashBPM,
+	}
+	kmsigversion = CBnTTest{
+		Name:     "KM Signature Version",
+		function: KMStructureValidSignatureVersion,
+	}
+	kmsigkey = CBnTTest{
+		Name:     "KM Signature Key",
+		function: KMStructureValidSignatureKey,
+	}
+	kmsigscheme = CBnTTest{
+		Name:     "KM Signature Scheme",
+		function: KMStructureValidSignatureSigSchemeCheck,
+	}
+	kmsigkeysize = CBnTTest{
+		Name:     "KM Signature Key Size",
+		function: KMStructureValidSigKeySize,
+	}
+	kmsighashalg = CBnTTest{
+		Name:     "KM Signature Hash Alg",
+		function: KMStructureValidSigHashAlg,
+	}
+	kmsigdata = CBnTTest{
+		Name:     "KM Signature Data",
+		function: KMStructureValidSigData,
+	}
+)
+
+// ImageTests list all CBnTTests in form of a slice
+var ImageTests = []*CBnTTest{
+	&hasFitEntries,
+	&kmValidHeader,
+	&kmValidKMSigOffset,
+	&kmValidReserverd2,
+	&kmrevisionvalid,
+	&kmsvnvalid,
+	&kmidvalid,
+	&kmpubkeyhash,
+	&kmkeyhashcount,
+	&kmbpmkeyhash,
+	&kmsigversion,
+	&kmsigkey,
+	&kmsigscheme,
+	&kmsigkeysize,
+	&kmsighashalg,
+	&kmsigdata,
+}
+
+// HasValidFitEntries defines the behavior for the CBnTTest "HasFitEntries"
+func HasValidFitEntries() (bool, error) {
+	var found int
+	fitentries, err := fit.GetTable(image)
+	if err != nil {
+		return false, err
+	}
+	for _, item := range fitentries {
+		if item.Type() == fit.EntryTypeBootPolicyManifest {
+			found++
+		}
+		if item.Type() == fit.EntryTypeKeyManifestRecord {
+			found += 2
+		}
+	}
+	switch found {
+	case 0:
+		// No KM/BPM
+		return false, fmt.Errorf("No KM and no BPM entry in FIT found")
+	case 1:
+		// NO BPM
+		return false, fmt.Errorf("No BPM entry in FIT found")
+	case 2:
+		// No KM
+		return false, fmt.Errorf("No KM entry in FIT found")
+	case 3:
+		// Success
+		return true, nil
+	default:
+		return false, fmt.Errorf("unexpected Error")
```

Comment:
Add a test if there are two keymanifestrecords or two bootpolicymanifests as well?

Reply:
Multiple KMs BPMs are not supported in general. Good idea for an issue.